xls/modules/zstd/README.md

```diff
@@ -0,0 +1,251 @@
+# ZSTD decoder
+
+The ZSTD decoder decompresses the correctly formed ZSTD frames and blocks.
+It implements the [RFC 8878](https://www.rfc-editor.org/rfc/rfc8878.html) decompression algorithm.
+Overview of the decoder architecture is presented on the diagram below.
+The decoder comprises:
+* frame decoder,
+* block dispatcher,
+* 3 types of processing units: RAW, RLE, and compressed,
+* command aggregator,
+* history buffer,
+* repacketizer.
+
+Incoming ZSTD frames are processed in the following order:
+1. magic number is detected,
+2. frame header is parsed,
+3. ZSTD data blocks are being redirected to correct processing unit based on the block header,
+4. processing unit results are aggregated in correct order into a stream
+and routed to the history buffer,
+5. data block outputs are assembled based on the history buffer contents and update history,
+6. decoded data is processed by repacketizer in order to prepare the final output of the decoder,
+7. (optional) calculated checksum is compared against frame checksum.
+
+![](img/ZSTD_decoder.png)
+
+## ZSTD decoder architecture
+
+### Top level Proc
+This state machine is responsible for receiving encoded ZSTD frames, buffering the input and passing it to decoder's internal components based on the state of the proc.
+The states defined for the processing of ZSTD frame are as follows:
+* DECODE_MAGIC_NUMBER
+* DECODE_FRAME_HEADER
+* DECODE_BLOCK_HEADER
+* FEED_BLOCK_DECODER
+* DECODE_CHECKSUM
+
+After going through initial stages of decoding magic number and frame header, decoder starts the block division process.
+It decodes block headers to calculate how many bytes must be sent to the block dispatcher and when the current frame's last data block is being processed.
+Knowing that it starts feeding the block decoder with data required for decoding current block.
+For this task it uses an 8 byte native interface: a 64-bit data bus and a 64-bit length field that contains the number of correct bits on the data bus.
+After transmitting all data required for current block, it loops around to the block header decoding state and when next block header is not found it decodes checksum when it was requested in frame header or finishes ZSTD frame decoding and loops around to magic number decoding.
+
+### ZSTD frame header decoder
+This part of the design starts with detecting the ZSTD magic number.
```

Comment:
s/This part of the design/This module/ ? to match with other description?

Reply:
Frame header decoding is actually implemented as a series of functions that are called in the top level proc. I'm not sure if `module` would be a good word here.